File: addiction_recovery_ai/utils/math_helpers.py
"""
Mathematical utility functions
"""

from typing import List


def calculate_percentage(part: float, total: float) -> float:
    """
    Calculate percentage
    
    Args:
        part: Part value
        total: Total value
    
    Returns:
        Percentage (0-100)
    """
    if total == 0:
        return 0.0
    
    return (part / total) * 100.0


def calculate_average(values: List[float]) -> float:
    """
    Calculate average of values
    
    Args:
        values: List of numeric values
    
    Returns:
        Average value
    """
    if not values:
        return 0.0
    
    return sum(values) / len(values)


def calculate_median(values: List[float]) -> float:
    """
    Calculate median of values
    
    Args:
        values: List of numeric values
    
    Returns:
        Median value
    """
    if not values:
        return 0.0
    
    sorted_values = sorted(values)
    length = len(sorted_values)
    
    if length % 2 == 0:
        mid = length // 2
        return (sorted_values[mid - 1] + sorted_values[mid]) / 2.0
    
    return sorted_values[length // 2]


def clamp_value(value: float, min_value: float, max_value: float) -> float:
    """
    Clamp value between min and max
    
    Args:
        value: Value to clamp
        min_value: Minimum allowed value
        max_value: Maximum allowed value
    
    Returns:
        Clamped value
    """
    return max(min_value, min(max_value, value))


def round_to_decimals(value: float, decimals: int = 2) -> float:
    """
    Round value to specified decimal places
    
    Args:
        value: Value to round
        decimals: Number of decimal places
    
    Returns:
        Rounded value
    """
    return round(value, decimals)

File: music_analyzer_ai/mobile_app/src/utils/math-helpers.ts
/**
 * Math utility functions
 * Common mathematical operations
 */

/**
 * Clamp number between min and max
 */
export function clamp(value: number, min: number, max: number): number {
  return Math.min(Math.max(value, min), max);
}

/**
 * Linear interpolation
 */
export function lerp(start: number, end: number, t: number): number {
  return start + (end - start) * t;
}

/**
 * Map value from one range to another
 */
export function mapRange(
  value: number,
  inMin: number,
  inMax: number,
  outMin: number,
  outMax: number
): number {
  return ((value - inMin) * (outMax - outMin)) / (inMax - inMin) + outMin;
}

/**
 * Round to decimal places
 */
export function roundTo(value: number, decimals: number): number {
  const factor = Math.pow(10, decimals);
  return Math.round(value * factor) / factor;
}

/**
 * Check if number is in range
 */
export function isInRange(value: number, min: number, max: number): boolean {
  return value >= min && value <= max;
}

/**
 * Get percentage of value in range
 */
export function getPercentage(value: number, min: number, max: number): number {
  if (max === min) return 0;
  return ((value - min) / (max - min)) * 100;
}

/**
 * Get value from percentage in range
 */
export function fromPercentage(
  percentage: number,
  min: number,
  max: number
): number {
  return min + (percentage / 100) * (max - min);
}

/**
 * Calculate distance between two points
 */
export function distance(
  x1: number,
  y1: number,
  x2: number,
  y2: number
): number {
  return Math.sqrt(Math.pow(x2 - x1, 2) + Math.pow(y2 - y1, 2));
}

/**
 * Calculate average of numbers
 */
export function average(numbers: number[]): number {
  if (numbers.length === 0) return 0;
  return numbers.reduce((sum, num) => sum + num, 0) / numbers.length;
}

/**
 * Calculate sum of numbers
 */
export function sum(numbers: number[]): number {
  return numbers.reduce((sum, num) => sum + num, 0);
}

/**
 * Get minimum value
 */
export function min(numbers: number[]): number {
  return Math.min(...numbers);
}

/**
 * Get maximum value
 */
export function max(numbers: number[]): number {
  return Math.max(...numbers);
}

/**
 * Calculate median
 */
export function median(numbers: number[]): number {
  const sorted = [...numbers].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0
    ? (sorted[mid - 1] + sorted[mid]) / 2
    : sorted[mid];
}

